bundles/io/org.eclipse.smarthome.io.net/src/main/java/org/eclipse/smarthome/io/net/http/internal/SecureHttpClientFactory.java

```diff
@@ -0,0 +1,219 @@
+        minThreadsShared = (int) parameters.get(CONFIG_MIN_THREADS_SHARED);
+        maxThreadsShared = (int) parameters.get(CONFIG_MAX_THREADS_SHARED);
+        keepAliveTimeoutShared = (int) parameters.get(CONFIG_KEEP_ALIVE_SHARED);
+        minThreadsCustom = (int) parameters.get(CONFIG_MIN_THREADS_CUSTOM);
+        maxThreadsCustom = (int) parameters.get(CONFIG_MAX_THREADS_CUSTOM);
+        keepAliveTimeoutCustom = (int) parameters.get(CONFIG_KEEP_ALIVE_CUSTOM);
+    }
+
+    private HttpClient createHttpClientInternal(String consumerName, String endpoint) {
+        logger.info("creating httpClient for endpoint {}", endpoint);
+        SslContextFactory sslContextFactory = new SslContextFactory();
+        sslContextFactory.setEndpointIdentificationAlgorithm("HTTPS");
+        if (endpoint != null && trustmanagerProvider != null) {
+            Stream<TrustManager> trustManagerStream = trustmanagerProvider.getTrustManagers(endpoint);
+            TrustManager[] trustManagers = trustManagerStream.toArray(TrustManager[]::new);
+            if (trustManagers.length > 0) {
+                logger.info("using custom trustmanagers (certificate pinning) for httpClient for endpoint {}",
+                        endpoint);
+                try {
+                    SSLContext sslContext = SSLContext.getInstance("TLS");
+                    sslContext.init(null, trustManagers, null);
+                    sslContextFactory.setSslContext(sslContext);
+                } catch (Exception ex) {
+                    throw new RuntimeException("Cannot create an TLS context for the endpoint '" + endpoint + "'!", ex);
+                }
+            }
+        }
+        String excludeCipherSuites[] = { "^.*_(MD5)$" };
+        sslContextFactory.setExcludeCipherSuites(excludeCipherSuites);
+
+        HttpClient httpClient = new HttpClient(sslContextFactory);
+        final ThreadPoolExecutor threadPoolExecutor = createThreadPool(consumerName, minThreadsCustom, maxThreadsCustom,
+                keepAliveTimeoutCustom);
+
+        // we need to add this thread pool to the client as managed object,
+        // so that it will become shutdown when the client is stopped
+        httpClient.addManaged(new AbstractLifeCycle() {
+            @Override
+            protected void doStop() throws Exception {
+                threadPoolExecutor.shutdown();
+            }
+        });
+
+        httpClient.setExecutor(threadPoolExecutor);
+
+        try {
+            httpClient.start();
+        } catch (Exception e) {
+            logger.error("Could not start jetty client", e);
+            throw new RuntimeException("Could not start jetty client", e);
+        }
+
+        return httpClient;
+    }
+
+    private ThreadPoolExecutor createThreadPool(String consumerName, int minThreads, int maxThreads,
+            int keepAliveTimeout) {
+        // We need an "empty" queue, because otherwise jetty workers become queued instead of executed.
```

Comment:
As discussed, we have in ESH the very powerful [QueueingThreadPoolExecutor](https://github.com/eclipse/smarthome/blob/master/bundles/core/org.eclipse.smarthome.core/src/main/java/org/eclipse/smarthome/core/common/QueueingThreadPoolExecutor.java), which imho has a much better behavior than a simple one that uses a SynchronousQueue - you should definitely look into using it.

Reply:
The jetty client uses the pool to create it's own worker threads. That's some kind of misuse, but we have to stick with it. Any kind of queuing makes no sense for the jetty client, so I would like to go with the non-queuing thread pool. If threads becomes exhausted, it's better to get an rejected-exception than random dead-lock behaviour.

Reply:
Can you prove that all tasks that the client submits are permanent workers?
This would mean that every client instance always uses exactly 4 threads, not more and not less. I'd love to see some proof for this, before I can believe it...

Reply:
4 is the minimal number of threads that become started per client. More worker threads are started and stopped on demand. But this should be under control of the jetty client. If threads become queued they are not started, but jetty will assume, that they were started. I think this is the reason for the deadlocks we have seen. If you like I can try it.

Reply:
> More worker threads are started and stopped on demand.

Which demand? What are they used for and what makes them being stopped again? If they only exist during a specific HTTP call, queuing tasks for a subsequent HTTP call should result in what we actually expect: That the number of concurrent calls is limited and the next is blocked until a previous is completed.

Reply:
The jetty client itself manages the worker threads. For example while testing with 90 parallel requests, the jetty client created around 40 threads, which disappeared after some time.
The whole lifecycle of the threads is managed by the jetty client. Introducing a separate queueing in the Executor would interfere with the jetty thread management.

Reply:
> The whole lifecycle of the threads is managed by the jetty client.

If this is the case, I do not see why we should at all interfere with it without running risk of breaking something.
The safest way would be to leave the Jetty implementation as it is, i.e. using the [QueuedThreadPool](https://github.com/eclipse/jetty.project/blob/jetty-9.3.x/jetty-util/src/main/java/org/eclipse/jetty/util/thread/QueuedThreadPool.java), which also offers setting minThreads, maxThreads, idleTimeout and a name - so all customisations that you require seem to be possible with it.

Reply:
The QueuedThreadPool does not support custom name for the threads. Maybe we can skip this requirement, as it is only useful for the customer clients.

Reply:
Hm, as mentioned above, [I thought it does](https://github.com/eclipse/jetty.project/blob/jetty-9.3.x/jetty-util/src/main/java/org/eclipse/jetty/util/thread/QueuedThreadPool.java#L251-L254)...

Reply:
Sorry, I didn't saw that. I changed the code to use the QueuedThreadPool.